#8108 bug in projection manager for dupes from group filters

jimasp opened the pull request.

sebastienros committed:
Fixing build.cmd

Commit:
#8108 bug in projection manager for dupes from group filters

Commit:
8108 - add ListIds() to IHQLQuery

Commit:
#8108 - add ListIds() to IHQLQuery

Commit:
Merge branch '8108' of https://github.com/jimasp/Orchard into 8108

Commit:
Merge branch 'dev' into 8108

Commit:
#8108 tweak for better single group performance

Commit:
Merge branch '8108' of https://github.com/jimasp/Orchard into 8108

sebastienros commented:
Can you give us some number for perf before and after the change on a scenario that is impacted by this change? And also on a scenario that shout not be impacted (no groups).

jimasp commented:
I ran a very quick + basic test.

I created a query based with Content Type "Any" as a filter.
I also created a Group query with Content Type "Any" OR Content Type "Any".

I then put a StopWatch start in ProjectionPartDriver like so:
```
  var metaData = Services.ContentManager.GetItemMetadata(part.ContentItem);
  Stopwatch stopwatch = new Stopwatch();
  stopwatch.Start();
```

And added Logger code to output the Elapsed.TotalMilliSeconds:
```
if (part.Record.DisplayPager) {
    var contentItemsCount = _projectionManager.GetCount(query.Id, part) - part.Record.Skip;
    contentItemsCount = Math.Max(0, contentItemsCount);
    pagerShape.TotalItemCount(contentItemsCount);
}
Logger.Debug("ProjectionDriver Elapsed:" + stopwatch.Elapsed.TotalMilliseconds);
```

Such that both `_projectionManager.GetContentItems()` and `_projectionManager.GetCount(query.Id, part)` are contained in the timer.

I disabled SysCache, and linked up a projection page with a pager to the test, with the setting for "Maximum number of items which can be queried at once" set to unlimited, and 10 items per page.

I loaded up an unrelated page in the site to start up Orchard and removed all scheduled tasks so unrelated db operations did not interfere with the test.
Then I loaded up the Projection Page 20 times with Ctrl+F5.

Tests run with 18365 (published) content items in the database:

With PR changes:
With Group: mean average - 120ms (range: 115 - 127ms)
Without Group: mean average - 97ms (range: 95 - 132ms)

Without PR changes:
With Group: mean average - 105ms (range: 98 - 148ms)
Without Group: mean average - 85ms (range: 81 - 90ms)

These times are from a debug build running on a fast dev PC.

sebastienros commented:
@BenedekFarkas what did we decide the next step is with this PR? Did you want to test it?

BenedekFarkas commented:
Yes, sorry, it's on my list, but haven't had time to do it yet.

BenedekFarkas commented on src/Orchard.Web/Modules/Orchard.Projections/Services/ProjectionManager.cs:
```diff
@@ -120,9 +120,12 @@ public class ProjectionManager : IProjectionManagerExtension {
+            if (queryRecord.FilterGroups.Count <= 1)
+                return contentQueries.Sum(contentQuery => contentQuery.Count());
+            else 
+                return contentQueries.SelectMany(contentQuery => contentQuery.ListIds()).Distinct().Count();
```
I have a fetish for the ternary-operator, to be honest, but besides that, I think this would be a bit cleaner:

```suggestion
            return queryRecord.FilterGroups.Count > 1 ?
                contentQueries.SelectMany(contentQuery => contentQuery.ListIds()).Distinct().Count() :
                contentQueries.Sum(contentQuery => contentQuery.Count());
```

I added this comment with the suggestion feature for the better syntax highlighting, but you won't be able to correctly apply it directly, because this change corresponds to multiple lines in your code.

BenedekFarkas replied:
@jimasp I added a suggestion for each (4) line involved, so you can commit them in a single batch.

BenedekFarkas commented on src/Orchard.Web/Modules/Orchard.Projections/Services/ProjectionManager.cs:
```diff
@@ -120,9 +120,12 @@ public class ProjectionManager : IProjectionManagerExtension {
-                .Sum(contentQuery => contentQuery.Count());
+            var contentQueries = GetContentQueries(queryRecord, Enumerable.Empty<SortCriterionRecord>(), tokens);
+
+            if (queryRecord.FilterGroups.Count <= 1)
```
```suggestion
```

BenedekFarkas commented on src/Orchard.Web/Modules/Orchard.Projections/Services/ProjectionManager.cs:
```diff
@@ -120,9 +120,12 @@ public class ProjectionManager : IProjectionManagerExtension {
+            var contentQueries = GetContentQueries(queryRecord, Enumerable.Empty<SortCriterionRecord>(), tokens);
+
+            if (queryRecord.FilterGroups.Count <= 1)
+                return contentQueries.Sum(contentQuery => contentQuery.Count());
```
```suggestion
```

BenedekFarkas commented on src/Orchard.Web/Modules/Orchard.Projections/Services/ProjectionManager.cs:
```diff
@@ -120,9 +120,12 @@ public class ProjectionManager : IProjectionManagerExtension {
+
+            if (queryRecord.FilterGroups.Count <= 1)
+                return contentQueries.Sum(contentQuery => contentQuery.Count());
+            else 
```
```suggestion
```

BenedekFarkas committed:
Merge branch 'dev' into 8108

BenedekFarkas committed:
Code styling in ProjectionManager

BenedekFarkas commented:
@sebastienros
I'm closing this: @jimasp's performance numbers are really promising and the change isn't huge. I also tested it with a site that has multiple thousands of published content items (the client is already running this code in production). I had some code styling suggestions which I committed to Jim's branch, so it's included in the PR.
